### ISSUE TITLE
Epinio API uses forward-auth authentication

### ISSUE THREAD

**Issue by jimmykarily** [epinio/epinio#667]
Follow up of #486 
Depends on #666 

## Acceptance Criteria

- Basic auth flow is completely removed
- Epinio is using forward-auth with dex and a static user for authentication
- The Epinio cli can still authenticate with the API
- The documentation is update and describes how forward auth works in Epinio (any images are updated too)

## Notes

Epinio cli should be able to authenticate and use the Epinio API. This can happens either completely automatically (by using the username/password to complete the oauth2 flow) or semi-automatically (by letting the user complete the flow and writing the final token in the config). If we automate the oauth2 flow in the epinio cli for the default static user, we should make sure we **don't** do that for other idps (e.g. GitHub). The whole purpose of using oauth2 is for the user to not have to give you their third party credentials. The static user credentials belong to Epinio so that's ok in that case.

**Comment by jimmykarily** [comment]
This is how some people have solved the programmatic authentication with dex: https://github.com/kubeflow/kfctl/issues/140#issuecomment-718512435

It's either this, or we can simply have an endpoint in epinio that shows the right epinio config section for the current token. So the user runs something like `epinio login` which gives url to the user to visit. Something like: 


```
$ epinio login
Visit: https://epinio.mydomain.org/cli-token
```
That will lead the user to the start of the oauth flow (the endpoint will be protected by the traefik-forward-auth middleware). When authentication is successul, the served page will say something like:

```
Add this section to your epinio config to allow the epinio cli to talk to Epinio:

token:
  ... things here ...

```

As an improvement, the cli could start a local server (like other clis do, e.g. `fly`). Then the original request will redirect to that local server passing the token in params. The cli will automatically write the config file then.

**Comment by jimmykarily** [comment]
> As an improvement, the cli could start a local server (like other clis do, e.g. fly). Then the original request will redirect to that local server passing the token in params. The cli will automatically write the config file then.

like what is described here for example: https://developer.okta.com/blog/2018/07/16/oauth-2-command-line#the-trick-to-oauth-20-on-the-command-line

**Comment by jimmykarily** [comment]
There is a problem with logging out users. Traefik forward auth, doesn't require the user to pass the Bearer token every time. Instead it creates a cookie which allows authorization until it's expired:

- https://github.com/thomseddon/traefik-forward-auth/blob/c4317b7503fb0528d002eb1e5ee43c4a37f055d0/README.md#logging-out
- https://github.com/thomseddon/traefik-forward-auth/blob/c4317b7503fb0528d002eb1e5ee43c4a37f055d0/internal/server.go#L95

This means, the only way for the user to be logged out, is to delete the cookie from the browser. There doesn't seem to be a way to block the user from accessing the protected Ingresses unless they want to.

**Comment by jimmykarily** [comment]
Working branch: `667-use-forward-auth`

**Comment by agracey** [comment]
Here's how I used forward auth in a previous project: https://github.com/agracey/code_server_k8s_sso/blob/master/chart/code-server-k8s/templates/portal_ingress.yaml

**Comment by jimmykarily** [comment]
Putting back in "Icebox" in favor of #686. We need to understand things better before we decide on a solution.

**Comment by enrichman** [comment]
Superseded by https://github.com/epinio/epinio/issues/1571. Closing.
